test/state/rlp.hpp

```diff
@@ -0,0 +1,117 @@
+// evmone: Fast Ethereum Virtual Machine implementation
+// Copyright 2021 The evmone Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <intx/intx.hpp>
+#include <cassert>
+#include <string>
+#include <string_view>
+#include <utility>
+#include <vector>
+
+namespace evmone::rlp
+{
+using bytes = std::basic_string<uint8_t>;
+using bytes_view = std::basic_string_view<uint8_t>;
+
+namespace internal
+{
+template <uint8_t ShortBase, uint8_t LongBase>
+inline bytes encode_length(size_t l)
+{
+    static constexpr auto short_cutoff = 55;
+    static_assert(ShortBase + short_cutoff <= 0xff);
+    assert(l <= 0xffffff);
+
+    if (l <= short_cutoff)
+        return {static_cast<uint8_t>(ShortBase + l)};
+    else if (const auto l0 = static_cast<uint8_t>(l); l <= 0xff)
+        return {LongBase + 1, l0};
+    else if (const auto l1 = static_cast<uint8_t>(l >> 8); l <= 0xffff)
+        return {LongBase + 2, l1, l0};
+    else
+        return {LongBase + 3, static_cast<uint8_t>(l >> 16), l1, l0};
+}
+
+inline bytes wrap_list(const bytes& content)
+{
+    return internal::encode_length<0xc0, 0xf7>(content.size()) + content;
+}
+
+template <typename InputIterator>
+inline bytes encode_container(InputIterator begin, InputIterator end);
+}  // namespace internal
+
+inline bytes_view trim(bytes_view b) noexcept
+{
+    b.remove_prefix(std::min(b.find_first_not_of(uint8_t{0x00}), b.size()));
+    return b;
+}
+
+template <typename T>
+inline decltype(rlp_encode(std::declval<T>())) encode(const T& v)
+{
+    return rlp_encode(v);
```

Comment:
confused about where is this defined

Reply:
You need to provide an overloading for any custom type.

Reply:
Doesn it not work to require to define `encode(const T& v)` overload directly?

Reply:
Probably, but there are 2 reasons to use different name:
1. These are ADL-aware functions defined outside of RLP namespace, so `encode()` is ambiguous.
2. Having single overloaded name for everything already caused issues, that's why I needed `tuple()` separated.